core/src/error.rs

```diff
@@ -0,0 +1,8 @@
+#[derive(Debug, Default)]
+pub struct CriticalError {}
```

Comment:
I think it makes sense to have some sort of inner type of error inside or message at least. otherwise it's hard to debug where're the problems

Reply:
I did consider adding a message, the thing is... I was thinking on using this more like a tag to be used with anyhow. The thing is, if you use the debug printing for the error with the current setup, it will look something like this:

```
Caused by:
    0: critical level error
    1: failed to deserialize block
```

Because all things you attach as a context are logged. It's still not ideal if you use the `Display` printing, but I think that's the case anyway with anyhow.

So, currently we could add a message, it's just that you can already attach one with a `context` anyway...

What I think may be worth it though, is something like this:

```rust
#[derive(Debug, Default)]
pub struct CriticalError {
    pub file: &'static str,
    pub line: u32,
}

impl std::fmt::Display for CriticalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "critical level error in {}:{}", self.file, self.line)
    }
}

#[macro_export]
macro_rules! critical_error {
    () => {
        CriticalError {
            line: line!(),
            file: file!(),
        }
    };
}
```

So we could get more information about the where.

Reply:
i like the idea to add file and line details to critical error, makes sense

Reply:
regarding the the message -- agree with your point